zookeeper/settings.sls

```diff
@@ -67,54 +67,47 @@
 {%- set hosts_target         = g.get('hosts_target', p.get('hosts_target', 'roles:zookeeper')) %}
 {%- set targeting_method     = g.get('targeting_method', p.get('targeting_method', 'grain')) %}
 
-{%- set force_mine_update    = salt['mine.send'](hosts_function) %}
-{%- set zookeepers_host_dict = salt['mine.get'](hosts_target, hosts_function, targeting_method) %}
-{%- set zookeepers_ids       = zookeepers_host_dict.keys() %}
-{%- set zookeepers_hosts     = zookeepers_host_dict.values() %}
-
-{%- set zookeeper_host_num   = zookeepers_ids | length() %}
-
-{%- if zookeeper_host_num == 0
-    or zookeeper_host_num is odd %}
-  # 0 means Zookeeper nodes have not been found,
-  # for 1, 3, 5 ... nodes just return the list
-  {%- set node_count = zookeeper_host_num %}
-{%- elif zookeeper_host_num is even %}
-  # for 2, 4, 6 ... nodes return (n -1)
-  {%- set node_count = zookeeper_host_num - 1 %}
+{%- set zookeepers_with_ids = [] %}
+{%- set zookeepers          = [] %}
+{%- set myid_dist           = [] %}
+{%- set connection_string   = [] %}
+{%- set minion_ips          = salt['network.ip_addrs']() %}
+
+{% if p.get('nodes') %}
+  {%- set zookeeper_nodes   = p.get('nodes', []) %}
+{%- else %}
+  {%- set force_mine_update = salt['mine.send'](hosts_function) %}
+  {%- set zookeeper_nodes   = salt['mine.get'](hosts_target, hosts_function, targeting_method) %}
 {%- endif %}
 
-# yes, this is not pretty, but produces sth like:
-# {'node1': '0+node1', 'node2': '1+node2', 'node3': '2+node2'}
-{%- set zookeepers_with_ids = {} %}
-{%- for i in range(node_count) %}
-  {%- if zookeepers_hosts[i] is string %}
-    # get plain value from Zookeeper hosts fetched by the Mine function
-    {%- set node_addr = zookeepers_hosts[i] %}
-  {%- elif zookeepers_hosts[i] is sequence %}
-    # if list provided, take the first value
-    {%- set node_addr = zookeepers_hosts[i] | first() %}
-  {%- else %}
-    # skip any other values including nested dicts
-    {%- set node_addr = '' %}
-  {%- endif %}
-
-  {%- if node_addr %}
-    {%- do zookeepers_with_ids.update({zookeepers_ids[i] : '{0:d}'.format(i) + '+' + node_addr })  %}
+{%- for node in zookeeper_nodes %}
+  {%- set node_id = loop.index %}
+  {%- set zookeeper_with_id = {"id": node_id, "address": node.encode('ascii')} %}
+  {%- do zookeepers_with_ids.append(zookeeper_with_id)  %}
+  {%- do connection_string.append( node.encode('ascii') + ':' + port | string() ) %}
+  {%- do zookeepers.append( node.encode('ascii') ) %}
```

Comment:
Hello @vutny ,
Unfortunately, I didn't find a solution how to include an address of a node in a json without encoding.
If I use `node | string` it adds unicode characters from python. Of course we can manually remove them, but it will look worse then encoding.
If you use `node | string` and look in your `/etc/zookeeper/conf/zoo.cfg` you will find that  addresses  of servers awe wrapped in quotes and before quotes there is `u` character. I.e.: 
`server.1=u'192.168.57.101':2888:3888`
Do you know how to properly include a string in jinja into a JSON without encoding?